pkg/cloud/sql/sql.go

```diff
@@ -236,16 +282,135 @@ func (l *LocalSqlServer) Query(ctx context.Context, connectionString string, que
+	}
+
+	if len(migrationImageContexts) > 0 {
+		updates, err := migrations.BuildMigrationImages(fs, migrationImageContexts)
+		if err != nil {
+			exit.GetExitService().Exit(fmt.Errorf("failed to build migration images: %w", err))
+		}
+
+		// wait for updates to complete
+		for update := range updates {
+			if update.Err != nil {
+				exit.GetExitService().Exit(fmt.Errorf("failed to build migration image: %w", update.Err))
+			}
+		}
+
+		// run the migrations
+		localMigrations := []migrations.LocalMigration{}
+
+		// Update the migration status
+		for dbName := range databasesToMigrate {
+			l.State[dbName].Status = string(DatabaseStatusApplyingMigrations)
+		}
+
+		l.Publish(l.State)
+
+		for dbName := range databasesToMigrate {
+			migration, ok := l.State[dbName]
+
+			if ok {
+				localMigrations = append(localMigrations, *migration.LocalMigration)
+			}
+		}
+
+		err = migrations.RunMigrations(localMigrations)
+		if err != nil {
+			exit.GetExitService().Exit(fmt.Errorf("failed to run migrations: %w", err))
+		}
+
+		// Update the status to running
+		for dbName := range databasesToMigrate {
+			l.State[dbName].Status = string(DatabaseStatusActive)
+		}
+
+		l.Publish(l.State)
+	}
+
+	return nil
+}
+
+func (l *LocalSqlServer) HandleUpdates(lrs resources.LocalResourcesState) {
+	databasesToMigrate := make(map[string]*resourcespb.SqlDatabaseResource)
+
+	// Check for new databases to migrate
+	for dbName, r := range lrs.SqlDatabases.GetAll() {
+		_, ok := l.State[dbName]
+
+		connectionString, err := l.ensureDatabaseExists(dbName)
+		if err != nil {
+			exit.GetExitService().Exit(fmt.Errorf("failed to ensure database exists: %w", err))
+		}
+
+		if !ok {
+			l.State[dbName] = &DatabaseState{
+				Status:           string(DatabaseStatusStarting),
+				ResourceRegister: r,
+				ConnectionString: connectionString,
+			}
+
+			l.Publish(l.State)
+		}
+
+		migrationPath := r.Resource.Migrations.GetMigrationsPath()
+
+		if migrationPath != "" && l.State[dbName].LocalMigration == nil {
+			dockerHost := "host.docker.internal"
+
+			if goruntime.GOOS == "linux" {
+				host := env.GetEnv("NITRIC_DOCKER_HOST", "172.17.0.1")
+
+				dockerHost = host.String()
+			}
+
+			l.State[dbName].Status = string(DatabaseStatusBuildingMigrations)
+			l.State[dbName].LocalMigration = &migrations.LocalMigration{
+				DatabaseName: dbName,
+				// Replace localhost with host.docker.internal to allow the container to connect to the host
+				ConnectionString: strings.Replace(connectionString, "localhost", dockerHost, 1),
+			}
+
+			databasesToMigrate[dbName] = r.Resource
+
+			l.Publish(l.State)
+		} else {
+			l.State[dbName].Status = string(DatabaseStatusActive)
+			l.Publish(l.State)
+		}
+	}
+
+	if len(databasesToMigrate) > 0 {
+		err := l.BuildAndRunMigrations(databasesToMigrate)
+		if err != nil {
+			exit.GetExitService().Exit(fmt.Errorf("failed to build and run migrations: %w", err))
+		}
+	}
```

Comment:
Including automated migrations in an update handle for local resource state might be a touch aggressive? If I'm reading this right, this will trigger migrations for existing and new databases on any resource state change (e.g. Adding a Topic).

For now I think providing a convenient way to manually trigger migrations should suffice, as you've already included in the dashboard.

Reply:
Updated